### PR TITLE
I19 rm getcontract func

### PR DESCRIPTION
**Pull request by sgoia**
branch ready to merge.

### CODE REVIEW

File: src/Web3Connection.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+	/****** */
+	/*** CORE */
+	/****** */
+
+	/**
+	 * @name start
+	 * @description Start the Application
+	 */
+	start = () => {
```
[the remark this answers is not in the document]
Comment (sgoia) [reply src/Web3Connection.js:69]: You are certainly right, upcoming code refactories could include some improvements on these ideas, there is always room for improvement, thank you.

**Review by moshmage — APPROVED** [review]
You can turn off `no-prototype-builtins` if you'd like. 

Other than that, lgtm :D

**Comment by Ruiub3i** [comment]
@sgoia Please also update the Readme so that people know how to use this new version of the architecture

**Comment by sgoia** [comment]
> You can turn off `no-prototype-builtins` if you'd like.
> 
> Other than that, lgtm :D

did it, thanks.

**Comment by Ruiub3i** [comment]
@sgoia is this one finalized?

**Comment by sgoia** [comment]
> @sgoia is this one finalized?

yes.

**Comment by Ruiub3i** [comment]
@sgoia Can you resolve the conflicts so that I can merge?

Thanks! :)

**Comment by sgoia** [comment]
> @sgoia Can you resolve the conflicts so that I can merge?
> 
> Thanks! :)

done.

**Comment by Ruiub3i** [comment]
@sgoia still has a few conflicts, all the rest looks ok :)

**Review by moshmage — CHANGES_REQUESTED** [review]
**please** remove `build` and `docs` folders and push the result

File: src/models/IContract.js [hunk 2]
```diff
@@ -15,32 +18,29 @@ import Contract from '../utils/Contract';
...
       this.params = {
-        web3,
+        web3Connection: this.web3Connection,
         abi,
         contractAddress,
         tokenAddress,
-        contract: new Contract(web3, abi, contractAddress),
       };
+
+      if (this.web3Connection.test) this._loadDataFromWeb3Connection();
```
Comment (moshmage) [inline src/models/IContract.js:43]: please don't add side effects to a constructor. if the user _needs_ this to happen, they will call it themselves.

File: src/models/IContract.js [hunk 3]
```diff
@@ -334,6 +323,77 @@ class IContract {
       throw new Error('Contract is paused');
     }
   }
+
+  /**
+   * @function
+   * @description Load data from Web3Connection object,
+   * Called at start when testing or at login on MAINNET
+   */
+  _loadDataFromWeb3Connection() {
```
Comment (moshmage) [inline src/models/IContract.js:332]: I'm not understanding this function:

Why would any of these values change and then need to be assigned again? Why don't we simply assign these on the constructor and be done with it for life?

File: tests/exchangeContract.js [hunk 4]
```diff
@@ -1,103 +1,105 @@
...
+    "should deploy Exchange Contract",
     mochaAsync(async () => {
       /* Create Contract */
-      exchangeContract = app.getExchangeContract({
-        contractAddress,
-      });
+      exchangeContract = new ExchangeContract({ contractAddress });
       /* Deploy */
-      const res = await exchangeContract.deploy({});
+      let res = await exchangeContract.deploy({});
```
Comment (moshmage) [inline tests/exchangeContract.js:40]: this is never reassigned, so it can be a `const`

Comment (sgoia) [reply tests/exchangeContract.js:40]: You are right, when testing only (not in API or lib) I do not use to pay critical attention to ret vs const as we could easily reassign adding some more asserts to the testing function and a change from const to let is needed.
We can certainly use const for sure if needed no problem.

File: src/models/IContract.js (hunk 3 above)
Comment (sgoia) [reply src/models/IContract.js:332]: A good example is "login" function, we are basically loading web3connection object, everything happens in it, so we usually need web3 object to connect to eth network and the account is we are using private key while testing.
***Mind that all those functions in Application.js like getStakingContract and getERC20Contract and similar, they are all used in a second step after instantiating Application and asserting it and then get our contracts objects using above functions, this issue was to solve exactly this, to remove that step, instantiate the contract we need without doing "app.getStakingContract()", this comes with a small price that the user does not need to know, we manage Web3Connection class to take care of everything else needed.

File: src/models/IContract.js [hunk 5]
```diff
@@ -15,32 +18,29 @@ import Contract from '../utils/Contract';
...
       if (!abi) {
         throw new Error('No ABI Interface provided');
       }
-      if (!web3) {
-        throw new Error('Please provide a valid web3 provider');
-      }
 
-      this.web3 = web3;
+      this.web3Connection = !web3Connection
+        ? new Web3Connection(params)
```
[the remark this answers is not in the document]
Comment (sgoia) [reply src/models/IContract.js:33]: No easy way to define in function docs, we basically create this web3Connection object from test and localtest params, everything old version of Application.js had is now in Web3Connection class, so we can either construct it from those params OR we can just pass an already created object and avoid instantiating a new one all the time. "if" conditions can be used no problem.

File: src/Web3Connection.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      );
+    }
+  }
+
+  /**
+   * Login with Metamask/Web3 Wallet - substitutes start()
+   * @function
+   * @return {Promise<boolean>}
+   */
+  login = async () => {
```
[the remark this answers is not in the document]
Comment (sgoia) [reply src/Web3Connection.js:106]: will update this.

**Review by moshmage — APPROVED** [review]

File: src/models/IContract.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (moshmage) [reply src/models/IContract.js]: erm, just do something like
```
@param {Web3} [web3connection=Web3Connection] if not provided will create a default one using provided values
```

File: src/models/IContract.js (hunk 2 above)
Comment (moshmage) [reply src/models/IContract.js:42]: introducing "if we are testing" code just makes us off-shot that much from the main implementation when testing, I'd rather we have to mimic the implementation when testing than having shortcuts tho

**Pull request merged**